pkg/asset/manifests/aws/zones.go

```diff
@@ -0,0 +1,272 @@
+// for managed or BYO VPC deployments from install-config.yaml.
+func setZones(in *zonesInput) error {
+	if len(in.InstallConfig.Config.AWS.Subnets) > 0 {
+		if err := in.GatherSubnetsFromMetadata(); err != nil {
+			return fmt.Errorf("failed to get subnets from metadata: %w", err)
+		}
+		return setZonesBYOVPC(in)
+	}
+
+	if err := in.GatherZonesFromMetadata(); err != nil {
+		return fmt.Errorf("failed to get availability zones from metadata: %w", err)
+	}
+	return setZonesManagedVPC(in)
+}
+
+// setZonesBYOVPC creates the CAPI NetworkSpec.Subnets setting the
+// desired subnets from install-config.yaml in the BYO VPC deployment.
+func setZonesBYOVPC(in *zonesInput) error {
+	if in.Cluster == nil {
+		return fmt.Errorf("failed to get AWSCluster config")
+	}
+	in.Cluster.Spec.NetworkSpec.VPC = capa.VPCSpec{
+		ID: in.Subnets.vpc,
+	}
+
+	for _, subnet := range in.Subnets.privateSubnets {
+		in.Cluster.Spec.NetworkSpec.Subnets = append(in.Cluster.Spec.NetworkSpec.Subnets, capa.SubnetSpec{
+			ID:               subnet.ID,
+			CidrBlock:        subnet.CIDR,
+			AvailabilityZone: subnet.Zone.Name,
+			IsPublic:         subnet.Public,
+		})
+	}
+
+	for _, subnet := range in.Subnets.publicSubnets {
+		in.Cluster.Spec.NetworkSpec.Subnets = append(in.Cluster.Spec.NetworkSpec.Subnets, capa.SubnetSpec{
+			ID:               subnet.ID,
+			CidrBlock:        subnet.CIDR,
+			AvailabilityZone: subnet.Zone.Name,
+			IsPublic:         subnet.Public,
+		})
+	}
+
+	return nil
+}
+
+// setZonesManagedVPC creates the CAPI NetworkSpec.VPC and the NetworkSpec.Subnets,
+// setting the desired zones from install-config.yaml in the managed
+// VPC deployment, when specified, otherwise default zones are discovered
+// from AWS API.
+// The CIDR blocks are calculated leaving free blocks to allow future expansions,
+// in Day-2, when desired.
+func setZonesManagedVPC(in *zonesInput) error {
+	out, err := extractZonesFromInstallConfig(in)
+	if err != nil {
+		return fmt.Errorf("failed to get availability zones: %w", err)
+	}
+
+	isPublishingExternal := in.InstallConfig.Config.Publish == types.ExternalPublishingStrategy
+	mainCIDR := capiutils.CIDRFromInstallConfig(in.InstallConfig)
+	in.Cluster.Spec.NetworkSpec.VPC = capa.VPCSpec{
+		CidrBlock: mainCIDR.String(),
+	}
+
+	// Base subnets considering only private zones, leaving one free block to allow
+	// future subnet expansions in Day-2.
+	numSubnets := len(out.AvailabilityZones) + 1
+
+	// Public subnets consumes one range from base blocks.
+	if isPublishingExternal {
+		numSubnets++
+	}
+
+	privateCIDRs, err := utilscidr.SplitIntoSubnetsIPv4(mainCIDR.String(), numSubnets)
+	if err != nil {
+		return fmt.Errorf("unable to retrieve CIDR blocks for all private subnets: %w", err)
+	}
+
+	var publicCIDRs []*net.IPNet
+	if isPublishingExternal {
+		publicCIDRs, err = utilscidr.SplitIntoSubnetsIPv4(privateCIDRs[len(out.AvailabilityZones)].String(), len(out.AvailabilityZones))
```

Comment:
Maybe lets add a comment here for future readers: 
```suggestion
        // The first num(zones) blocks are dedicated to the private subnets
		publicCIDRs, err = utilscidr.SplitIntoSubnetsIPv4(privateCIDRs[len(out.AvailabilityZones)].String(), len(out.AvailabilityZones))
```